### PR TITLE
detect server outtages, notify users

### PR DESCRIPTION
**Pull request by jacobheric**
Whenever a streaming connection dies, we now notify the frontend via web socket and show them a message.

Sorry this one got pretty big as I uncovered a lot of ways that those connection would die that wasn't getting handled properly. So I had to refactor our streams and sockets a bit. I also uncovered some other ways the console died ungracefully when the server dies and fixed those. 

Screenshot 2023-10-18 at 9 21 40 AM

### CODE REVIEW

File: components/error/internal.tsx [hunk 1]
```diff
@@ -0,0 +1,31 @@
+export const InternalError = ({ message }: { message: string }) => (
```
Comment (jacobheric) [inline components/error/internal.tsx:1]: This is a generic error component that we use on 500s.

File: components/error/server.tsx [hunk 2]
```diff
@@ -0,0 +1,31 @@
+export const ServerError = ({ message }: { message: string }) => (
```
Comment (jacobheric) [inline components/error/server.tsx:1]: I made the server error it's own thing as it's pretty special and we have to use some very heavy handed techniques to detect that our server streams are dead.

File: islands/opModal.tsx [hunk 3]
```diff
@@ -104,14 +86,18 @@ export default function OpModal(
...
     if (opModal.value) {
-      const schema = await getSchema();
-      opModal.value = {
-        ...opModal.value,
-        schemaInfo: {
-          schema: JSON.stringify(schema.schema, null, 2),
-          version: schema.version,
-        },
-      };
+      try {
```
Comment (jacobheric) [inline islands/opModal.tsx:89]: When the server died this fetch was blowing up the frontend.

File: lib/audienceMetrics.ts [hunk 4]
```diff
@@ -24,11 +25,13 @@ export const getAudienceMetrics = async ({ socket }: { socket: WebSocket }) => {
...
 
+    await audienceMetricsCall.headers;
+    serverErrorSignal.value = "";
+
     for await (const response of audienceMetricsCall?.responses) {
       for (const [key, value] of Object.entries(response?.rates)) {
-        !response?.Keepalive && socket.send(
-          JSON.stringify({ [key]: value as RateMetrics }),
-        );
+        !response?.Keepalive &&
```
Comment (jacobheric) [inline lib/audienceMetrics.ts:33]: Broke out socket send here to distinguish socket errors from server connection errors.

File: lib/stream.ts [hunk 5]
```diff
@@ -23,10 +24,15 @@ export const streamServiceMap = async () => {
     const call: ServerStreamingCall<GetAllRequest, GetAllResponse> = client
       .getAllStream({}, { ...meta, abort: abortController.signal });
 
-    void processResponses(call);
+    await call.headers;
```
Comment (jacobheric) [inline lib/stream.ts:27]: Clear errors on successful reconnect.

File: islands/serviceMap.tsx [hunk 6]
```diff
@@ -144,7 +152,9 @@ export default function ServiceMapComponent(
       const updated = updateNode(nodes, opUpdateSignal.value);
       setNodes(updated);
     }
+  });
 
+  useSignalEffect(() => {
```
Comment (jacobheric) [inline islands/serviceMap.tsx:157]: This was actually a fairly significant unrelated bug. We had one signal effect handler to track changes on both opModal and the service map...so we conflated those two side effects.

**Review by Adelle-Pitsas — APPROVED** [review]

**Pull request merged**